File: Week 2 - Algorithms/W2 Main Problems/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>

int lcount(string text);
int wcount(string text);
int scount(string text);

int main(void)
{
    // Prompt the user for some text
    string text = get_string("Text: \n");

    // int lcount = count letters
    int letters = lcount(text);

    // int wcount = count words
    int words = wcount(text);

    // int scount = count sentences
    int sentences = scount(text);

    // Coleman Liau Function

    // L = average number of letters per 100 words
    float L = ((float) letters / words) * 100.0;

    // S = average number of sentences per 100 words
    float S = ((float) sentences / words) * 100.0;

    // index = 0.0588 * L - 0.296 * S - 15.8
    float result = (0.0588 * L) - (0.296 * S) - 15.8;

    // Print the Grade Level
    if (result < 0)
    {
        printf("Before Grade 1\n");
    }
    else if (result >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %.0f\n", result);
    }
}

int lcount(string text)
{

    int n = strlen(text);
    int x = 0;

    for (int i = 0; i < n; i++)
    {
        if (isalpha(text[i]))
        {
            x += 1;
        }
    }

    return x;
}

int scount(string text)
{

    int n = strlen(text);
    int x = 0;

    for (int i = 0; i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            x += 1;
        }
    }

    return x;
}

int wcount(string text)
{

    int n = strlen(text);
    int x = 1;

    for (int i = 0; i < n; i++)
    {
        if (isspace(text[i]))
        {
            x += 1;
        }
    }

    return x;
}

File: W6_Python/W6_Problem_Sets/sentimental-readability/readability.py
# Import necessary modules
import re


def lcount(text):
    # Use list comprehension and isalpha() to count letters
    return sum(1 for char in text if char.isalpha())


def wcount(text):
    # Split the text by spaces and count the words
    return len(text.split())


def scount(text):
    # Use re.findall to count '.', '!', and '?'
    return len(re.findall(r'[.!?]', text))


def main():
    # Prompt the user for some text
    text = input("Text: \n")

    # Count letters, words, and sentences
    letters = lcount(text)
    words = wcount(text)
    sentences = scount(text)

    # Calculate L and S for the Coleman-Liau index
    L = (letters / words) * 100
    S = (sentences / words) * 100

    # Calculate the index
    result = 0.0588 * L - 0.296 * S - 15.8

    # Determine and print the grade level
    if result < 1:
        print("Before Grade 1")
    elif result >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(result)}")


# Call the main function
if __name__ == "__main__":
    main()
